### PR TITLE
Add raw delete range interface

### PR DESCRIPTION
**Pull request by MyonKeminta**
* Add raw delete range method to `Storage`
* Add RPC interface to service/kv.rs
* Add related metrics

### CODE REVIEW

**Comment by siddontang** [comment]
Please add test

File: src/server/service/kv.rs [hunk 1]
```diff
@@ -746,6 +748,48 @@ impl<T: RaftStoreRouter + 'static> tikvpb_grpc::Tikv for Service<T> {
...
+
+        let (cb, future) = make_callback();
+        let res = self.storage.async_raw_delete_range(
+            req.take_context(),
+            req.take_start_key(),
+            req.take_end_key(),
+            cb,
+        );
+        if let Err(e) = res {
+            self.send_fail_status(ctx, sink, Error::from(e), RpcStatusCode::ResourceExhausted); //???korewa
```
Comment (zhangjinpeng1987) [inline src/server/service/kv.rs]: What's the comment's meaning?

File: src/storage/mod.rs [hunk 2]
```diff
@@ -961,6 +961,36 @@ impl Storage {
...
     }
 
+    pub fn async_raw_delete_range(
+        &self,
+        ctx: Context,
+        start_key: Vec<u8>,
+        end_key: Vec<u8>,
+        callback: Callback<()>,
+    ) -> Result<()> {
+        for key in &[&start_key, &end_key] {
```
Comment (zhangjinpeng1987) [inline src/storage/mod.rs]: if start_key.len() > max || end_key.len() > max {
}

File: src/storage/mod.rs [hunk 3]
```diff
@@ -968,12 +968,16 @@ impl Storage {
...
-                callback(Err(Error::KeyTooLarge(key.len(), self.max_key_size)));
-                return Ok(());
-            }
+        if start_key.len() > self.max_key_size || end_key.len() > self.max_key_size {
+            let len = if start_key.len() > self.max_key_size {
+                start_key.len()
+            } else {
+                end_key.len()
+            };
+            callback(Err(Error::KeyTooLarge(len, self.max_key_size)));
```
Comment (zhangjinpeng1987) [inline src/storage/mod.rs]: cmp::max(start_key.len(), end_key.len())

**Comment by huachaohuang** [comment]
Please add test

**Comment by siddontang** [comment]
Great work!!!

File: src/storage/mod.rs [hunk 4]
```diff
@@ -1753,4 +1786,99 @@ mod tests {
         );
         storage.stop().unwrap();
     }
+
+    #[test]
+    fn test_raw_delete_range() {
+        let read_pool = ReadPool::new("readpool", &readpool::Config::default_for_test(), || {
```
Comment (siddontang) [inline src/storage/mod.rs]: I see too many places use the same way to create the test read pool, can we extract them to one simple function?

@breeswish

Comment (MyonKeminta) [reply src/storage/mod.rs]: let's do this in another PR?

Comment (breeswish) [reply src/storage/mod.rs]: agree. let's do it in another PR.

**Comment by zhangjinpeng1987** [comment]
LGTM please fix conflict.

**Comment by zhangjinpeng1987** [comment]
https://github.com/pingcap/kvproto/pull/241 have merged, please update kvproto.

File: src/storage/mod.rs [hunk 5]
```diff
@@ -1753,4 +1786,99 @@ mod tests {
...
+            read_pool_context_factory
+        });
+        let config = Config::default();
+        let mut storage = Storage::new(&config, read_pool).unwrap();
+        storage.start(&config).unwrap();
+        let (tx, rx) = channel();
+
+        // Write some key-value pairs to the db
+        for kv in &[
```
Comment (huachaohuang) [inline src/storage/mod.rs]: Assign these samples to a variable so that you don't need to list them below again.

**Comment by huachaohuang** [comment]
LGTM

**Pull request merged**